toolchain/check/node_block_stack.cpp

```diff
@@ -6,43 +6,64 @@
 
 #include "common/vlog.h"
 #include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringExtras.h"
 #include "toolchain/sem_ir/node.h"
 
 namespace Carbon::Check {
 
 auto NodeBlockStack::Push(SemIR::NodeBlockId id) -> void {
   CARBON_VLOG() << name_ << " Push " << stack_.size() << "\n";
-  CARBON_CHECK(stack_.size() < (1 << 20))
+  CARBON_CHECK(size() < (1 << 20))
       << "Excessive stack size: likely infinite loop";
-  stack_.push_back(id);
+  if (size_ == stack_.size()) {
+    stack_.emplace_back();
+  }
+  stack_[size_].Reset(id);
+  ++size_;
 }
 
 auto NodeBlockStack::PeekForAdd(int depth) -> SemIR::NodeBlockId {
-  CARBON_CHECK(static_cast<int>(stack_.size()) > depth) << "no such block";
-  int index = stack_.size() - depth - 1;
+  CARBON_CHECK(static_cast<int>(size()) > depth) << "no such block";
+  int index = size() - depth - 1;
   auto& slot = stack_[index];
-  if (!slot.is_valid()) {
-    slot = semantics_ir_->AddNodeBlock();
-    CARBON_VLOG() << name_ << " Add " << index << ": " << slot << "\n";
+  if (!slot.id.is_valid()) {
+    slot.id = semantics_ir_->AddNodeBlock();
+    CARBON_VLOG() << name_ << " Add " << index << ": " << slot.id << "\n";
   }
-  return slot;
+  return slot.id;
 }
 
 auto NodeBlockStack::Pop() -> SemIR::NodeBlockId {
-  CARBON_CHECK(!stack_.empty()) << "no current block";
-  auto back = stack_.pop_back_val();
-  CARBON_VLOG() << name_ << " Pop " << stack_.size() << ": " << back << "\n";
-  if (!back.is_valid()) {
+  CARBON_CHECK(!empty()) << "no current block";
+  --size_;
+  auto& back = stack_[size_];
+
+  // Finalize the block.
+  if (!back.content.empty() && back.id != SemIR::NodeBlockId::Unreachable) {
+    if (back.id.is_valid()) {
+      semantics_ir_->SetNodeBlock(back.id, back.content);
+    } else {
+      back.id = semantics_ir_->AddNodeBlock(back.content);
+    }
+  }
+
+  CARBON_VLOG() << name_ << " Pop " << size() << ": " << back.id << "\n";
+  if (!back.id.is_valid()) {
     return SemIR::NodeBlockId::Empty;
   }
-  return back;
+  return back.id;
 }
 
 auto NodeBlockStack::PrintForStackDump(llvm::raw_ostream& output) const
     -> void {
   output << name_ << ":\n";
-  for (auto [i, entry] : llvm::enumerate(stack_)) {
-    output << "\t" << i << ".\t" << entry << "\n";
+  for (const auto& [i, entry] : llvm::enumerate(stack_)) {
+    output << "\t" << i << ".\t" << entry.id << "\t{";
+    llvm::ListSeparator sep;
+    for (auto id : entry.content) {
```

Comment:
Just to note, this might not fit on one line (typically would not, I think). We should probably consider in general how much we're printing in these, but noting the stack would also be very large, that's nothing new to this PR so fine as-is.